// @devvie/bottom-sheet/lib/commonjs/components/defaultHandleBar/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DEFAULT_HANDLE_BAR_DEFAULT_HEIGHT = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
/**
 * This is the default handle bar component used when no custom handle bar component is provided
 */
const DefaultHandleBar = _ref => {
  let {
    style,
    ...otherProps
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, _extends({
    style: materialStyles.dragHandleContainer
  }, otherProps), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [materialStyles.dragHandle, style]
  }));
};
const materialStyles = _reactNative.StyleSheet.create({
  dragHandleContainer: {
    padding: 18,
    width: 50,
    alignSelf: 'center'
  },
  dragHandle: {
    height: 4,
    width: 32,
    backgroundColor: '#49454F',
    opacity: 0.4,
    alignSelf: 'center',
    borderRadius: 50
  }
});

// this will be used in `convertHeight` for clamping sheet height
const DEFAULT_HANDLE_BAR_DEFAULT_HEIGHT = exports.DEFAULT_HANDLE_BAR_DEFAULT_HEIGHT = 25; // paddingTop (10) + paddingBottom (10) + height (5)
var _default = exports.default = DefaultHandleBar;
